proof/courier.go

```diff
@@ -790,7 +915,157 @@ func (h *HashMailCourier) SetSubscribers(
 
 // A compile-time assertion to ensure the HashMailCourier meets the
 // proof.Courier interface.
-var _ Courier[Recipient] = (*HashMailCourier)(nil)
+var _ Courier = (*HashMailCourier)(nil)
+
+// UniverseRpcCourier is a universe RPC proof courier service handle. It
+// implements the Courier interface.
+type UniverseRpcCourier struct {
+	// cfg contains the courier's configuration parameters.
+	cfg *HashMailCourierCfg
+
+	// client is the RPC client that the courier will use to interact with
+	// the universe RPC server.
+	client unirpc.UniverseClient
+
+	// deliveryLog is the log that the courier will use to record the
+	// attempted delivery of proofs to the receiver.
+	deliveryLog DeliveryLog
+
+	// subscribers is a map of components that want to be notified on new
+	// events, keyed by their subscription ID.
+	subscribers map[uint64]*fn.EventReceiver[fn.Event]
+
+	// subscriberMtx guards the subscribers map and access to the
+	// subscriptionID.
+	subscriberMtx sync.Mutex
+}
+
+// DeliverProof attempts to delivery a proof to the receiver.
+func (c *UniverseRpcCourier) DeliverProof(ctx context.Context,
+	recipient Recipient, annotatedProof *AnnotatedProof) error {
+
+	// Construct universe key.
+	universeID := unirpc.ID{
+		Id: &unirpc.ID_AssetId{
+			AssetId: recipient.AssetID[:],
+		},
+	}
+
+	// Serialize the script key in compressed format. The RPC endpoint can
+	// handle a compressed script key.
+	scriptKeyBytes := recipient.ScriptKey.SerializeUncompressed()
+
+	outPoint := annotatedProof.AssetSnapshot.OutPoint
+
+	assetKey := unirpc.AssetKey{
+		Outpoint: &unirpc.AssetKey_Op{
+			Op: &unirpc.Outpoint{
+				HashStr: outPoint.Hash.String(),
+				Index:   int32(outPoint.Index),
+			},
+		},
+		ScriptKey: &unirpc.AssetKey_ScriptKeyBytes{
+			ScriptKeyBytes: scriptKeyBytes,
+		},
+	}
+
+	universeKey := unirpc.UniverseKey{
+		Id:      &universeID,
+		LeafKey: &assetKey,
+	}
+
+	// Construct asset leaf.
+	rpcAsset, err := taprpc.MarshalAsset(
+		ctx, annotatedProof.Asset, true, true, nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Extract the last proof from the annotated proof.
+	proofFile := &File{}
+	err = proofFile.Decode(bytes.NewReader(annotatedProof.Blob))
+	if err != nil {
+		return err
+	}
+	rawProof, err := proofFile.RawLastProof()
+	if err != nil {
+		return err
+	}
+
+	assetLeaf := unirpc.AssetLeaf{
+		Asset:         rpcAsset,
+		IssuanceProof: rawProof,
+	}
+
+	// Submit proof to courier.
+	_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
+		Key:       &universeKey,
+		AssetLeaf: &assetLeaf,
+	})
```

Comment:
Hmm, I'm not 100% sure how we need to do this. On the receiver side we expect the full proof file (containing the whole chain of proofs). But the receiver will only try to fetch that full proof file given the last proof's script key (at least that's how it works now with the hashmail courier).

So I think we might need to change the universe to also allow us to store and fetch whole files, not just single proofs. Not sure how that would look in the tree though.

Or, I guess the alternative would be to do it exactly as you propose, send each individual proof to the universe (that being a no-op for proofs that already exist), but then when _fetching_ a proof in the universe RPC courier we would need to assemble the whole proof chain/file again by iteratively querying each proof in the chain.

Not sure which path you had in mind, @Roasbeef.

Reply:
We could also reconstruct a proof file on the server side and offer that up via some RPC (perhaps the same RPC endpoints we currently have but pluralised?).

Reply:
> So I think we might need to change the universe to also allow us to store and fetch whole files, not just single proofs. Not sure how that would look in the tree though.

The way things are set up rn, you only store the key for _that_ state transition. We'll have a bit of a boostrap phase to start with, but you can imagine that in the future, just like a block explorer, you'll be able to easily look up w/e you need in the universe. Eventually, there'll be a large set of common prefixes (initial issuance, then all the UTXOs from that) that'll be widely cached/known. 

If we store the full proof file, then we'd give up the deduplication we get. Consider that if there was only a single UTXO that created an asset, then as you go on, all assets have a larger shared prefix. By only storing the state transitions in isolation, we can avoid needing to store that prefix over and over again. 

Going back to the top, assuming that we're starting from scratch, with a fresh universe server just spun up: yes the sender will submit each state transition they have individually. On the flip side, after the receiver gets that last state transition, then then crawl backwards via the pointers in the tree to assemble the proof back to genesis. You can think of it kind of like the popularized, but never implemented, "backwards header sync" of Bitcoin. You get the last header, then use the universe APIs to walk backwards.

Reply:
> We could also reconstruct a proof file on the server side and offer that up via some RPC (perhaps the same RPC endpoints we currently have but pluralised?).

What's the "server side" in this example?

Reply:
The universe server. So instead of the receiver needing to make multiple RPC requests to fetch the whole chain, there could be a new RPC that assembles the full chain on the universe server and sends it back as a full file to the client. But maybe that can be done as an optimization later on.